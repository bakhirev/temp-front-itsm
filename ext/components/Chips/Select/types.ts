export interface IItemChipSelect {
  /** Уникальный идентификатор чипса */
  id: string;
  /** Подпись к чипсу */
  label: string;
  /** Отключение чипса */
  disabled?: boolean;
  /** Иконка */
  icon?: any;
}
