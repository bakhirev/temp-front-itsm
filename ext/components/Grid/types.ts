import { ReactNode } from 'react';

/** Вид устройства/разрешения для задания свойств */
export type Device = 'mobile320' | 'tablet768' | 'tablet1024' | 'desktop1280' | 'desktop1600';

export type JustifyContent =
  | 'center'
  | 'flex-start'
  | 'flex-end'
  | 'space-around'
  | 'space-between';
export type AlignItems = 'baseline' | 'center' | 'flex-start' | 'flex-end' | 'stretch';

export const Devices: Device[] = [
  'mobile320',
  'tablet768',
  'tablet1024',
  'desktop1280',
  'desktop1600',
];

export interface IGridBaseProps {
  /** Элементы содержимого */
  children?: ReactNode;
  /** Имя класса для переопределения стилей */
  className?: string;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
}

/** Содержит размеры в количестве колонок, которое занимает данная ячейка на заданном устройстве/разрешении */
export interface IDevices {
  mobile320?: number;
  tablet768?: number;
  tablet1024?: number;
  desktop1280?: number;
  desktop1600?: number;
}

export interface IGridCellProps extends IGridBaseProps, IDevices {}

export interface IGridProps extends IGridBaseProps {
  /** Отключение отступов внутри модульной сетки */
  noindent?: boolean | undefined;
  /** Отключение отступов снаружи модульной сетки */
  nomargin?: boolean | undefined;
  /** Определяет распределение пространства у элементов контента вдоль главной оси flex контейнера */
  justifyContent?: JustifyContent;
  /** Выравнивает элементы текущей flex-линии аналогично justifyContent, но в перпендикулярном направлении */
  alignItems?: AlignItems;
}

export interface IGridContext {
  noindent?: boolean | undefined;
}
