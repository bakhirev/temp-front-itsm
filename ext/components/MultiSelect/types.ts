import type { Size } from './constants';

export interface IMultiSelectItem {
  label: string;
  value: string | number;
  disabled?: boolean;
}

export type IMultiSelectItemOptions = IMultiSelectItem | React.ReactNode;

export interface IInputProps {
  size: Size;
  disabled?: boolean;
  placeholder?: string;
  clearable?: boolean;
  menuIsOpen?: boolean;
  maxWidthChip?: string;
  currentMultiValue: IMultiSelectItemOptions[];
  onChange: (value: IMultiSelectItemOptions[]) => void;
}

export interface IChipItemProps {
  value: string | number;
  disabled?: boolean;
  maxWidthChip?: string;
  currentMultiValue: IMultiSelectItemOptions[];
  onChange: (value: IMultiSelectItemOptions[]) => void;
}

export interface IDropdownListProps {
  size: Size;
  disabled?: boolean;
  menuIsOpen: boolean;
  error?: boolean;
  classNameDropdownList?: string;
  dropdownListBorders: boolean;
  optionsList: IMultiSelectItemOptions[];
  currentMultiValue: IMultiSelectItemOptions[];
  setOpened: (menuIsOpen: any) => void;
  onChange: (value: IMultiSelectItemOptions[]) => void;
}
