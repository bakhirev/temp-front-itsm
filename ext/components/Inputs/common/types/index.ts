/** Типы для компонентов инпут */

export type Size = 'big' | 'small' | 'micro';
export type Status = 'success' | 'error' | 'default';
export type Mask = 'card' | 'phone';
export type Type = 'text' | 'password' | 'suggest' | 'number' | 'range' | 'informer';
