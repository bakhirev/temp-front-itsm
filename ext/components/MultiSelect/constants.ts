export type SelectType = 'default';
export type Size = 'big' | 'small' | 'micro';

export const VALUE_CONTAINER_MARGIN_RIGHT = '5px';
export const LABEL_MARGIN_BOTTOM = '8px';
export const ADDITIONAL_TEXT_MARGIN_TOP = '8px';
export const BOTTOM_LINE_HEIGHT_ON_FOCUS = '2px';
export const SELECT_DEFAULT_WIDTH = '320px';
export const CHIP_DEFAULT_MAX_WIDTH = '60px';
export const CLOSE_ICON_MARGIN_RIGHT = '4px';
export const ICON_DEFAULT_SIZE = 24;
export const ICON_MICRO_SIZE = 20;

export const CONTROL_HEIGHT = {
  micro: '32px',
  small: '40px',
  big: '56px',
};
export const OPTION_HEIGHT = {
  micro: '32px',
  small: '40px',
  big: '48px',
};

export const OPTION_PADDING = {
  micro: '6px 12px',
  small: '8px 16px',
  big: '12px 16px',
};
export const CONTROL_PADDING = (type = 'default') => ({
  micro: '6px 11px',
  small: '8px 15px',
  big: `${type === 'default' ? '12px 15px' : '10px 15px'}`,
});

export const MAX_LINES_AMOUNT_IN_MENU = 8;
export const MENU_VERTICAL_PADDING = '8px';
export const MENU_MAX_HEIGHT = {
  micro: `${
    parseInt(OPTION_HEIGHT.micro) * MAX_LINES_AMOUNT_IN_MENU + parseInt(MENU_VERTICAL_PADDING) * 2
  }px`,
  small: `${
    parseInt(OPTION_HEIGHT.small) * MAX_LINES_AMOUNT_IN_MENU + parseInt(MENU_VERTICAL_PADDING) * 2
  }px`,
  big: `${
    parseInt(OPTION_HEIGHT.big) * MAX_LINES_AMOUNT_IN_MENU + parseInt(MENU_VERTICAL_PADDING) * 2
  }px`,
};

export const MARK_HEIGHT_SMALL = '16px';
export const MARK_HEIGHT_BIG = '20px';
export const MARK_WIDTH_SMALL = '16px';
export const MARK_WIDTH_BIG = '20px';
export const CHECKBOX_MARGIN = '2px';
export const MARK_BEFORE_OFFSET = '-8px';

export const BORDER_RADIUS = '3px';
export const DROPDOWN_MARGIN_BOTTOM_TOP = '8px';
export const INPUT_BORDER_ACTIVE = '2px';
export const INPUT_BORDER_ERROR = '1px';
