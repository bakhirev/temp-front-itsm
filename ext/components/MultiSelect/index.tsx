import React, { ReactNode, FC, useState, useEffect, useCallback } from 'react';

import DropdownList from './components/DropdownList';
import Input from './components/Input';
import type { SelectType, Size } from './constants';
import {
  SELECT_DEFAULT_WIDTH,
  DROPDOWN_MARGIN_BOTTOM_TOP,
  INPUT_BORDER_ERROR,
  INPUT_BORDER_ACTIVE,
  CHIP_DEFAULT_MAX_WIDTH,
} from './constants';
import {
  SelectWrapper,
  Label,
  SelectContainer,
  AdditionalText,
} from './styled-components/otherComponents';
import { IMultiSelectItem } from './types';

export type IMultiSelectItemOptions = IMultiSelectItem | React.ReactNode;

export interface IMultiSelectProps {
  /** Ширина мульти селекта */
  width?: string;
  /** Список опций */
  list: IMultiSelectItemOptions[];
  /** Значение мульти селекта */
  value: IMultiSelectItemOptions[];
  /** Коллбэк на изменение значения multiSelect. MultiSelect является управляемым(controlled) компонентом, поэтому при срабатывании onChange требуется обновления пропса value */
  onChange: (value: IMultiSelectItemOptions[]) => void;
  /** Размер компонента */
  size: Size;
  /** Отключение компонента */
  disabled?: boolean;
  /** Селект с возможностью очищения */
  clearable?: boolean;
  /** Селект в состоянии ошибки */
  error?: boolean;
  /** Верхняя надпись к селекту */
  label?: ReactNode | string;
  /** Плейсхолдер селекта */
  placeholder?: string;
  /** Дополнительный текст под селектом */
  additionalText?: string;
  /** Тип селекта */
  type?: SelectType;
  /** Класс на SelectContainer блок */
  classNameSelectContainer?: string;
  /** Класс на Dropdown блок */
  classNameDropdownList?: string;
  /** Открыть и закрыть меню */
  menuOpenAndClose?: boolean;
  /** Имя класса для переопределения стилей */
  className?: string;
  /** Максимальная ширина чипсов */
  maxWidthChip?: string;
}

export const MultiSelect: FC<IMultiSelectProps> = (props) => {
  const {
    placeholder = '',
    label,
    disabled,
    clearable,
    list,
    value,
    onChange,
    size = 'big',
    additionalText,
    error = false,
    width = SELECT_DEFAULT_WIDTH,
    type = 'default',
    classNameSelectContainer = '',
    classNameDropdownList = '',
    menuOpenAndClose = false,
    className,
    maxWidthChip = CHIP_DEFAULT_MAX_WIDTH,
  } = props;

  // Убираем дублирование данных
  const optionsList = list.reduce(
    (st: any, el: any) =>
      st.map((e: { value: any }) => e.value).includes(el.value) ? st : [...st, el],
    []
  );

  // Дропдаун открыт/закрыт
  const [menuIsOpen, setOpened] = useState<boolean>(!!menuOpenAndClose);

  const [dropdownListBorders, setDropdownListMoveBorders] = useState<boolean>(false);

  useEffect(() => {
    !!menuOpenAndClose && setOpened(menuOpenAndClose);
  });

  const handleSelectWrapperClick = useCallback(() => {
    // При открытии выпадающий список не вылезает за окно браузера
    const optionWrapperDiv = document.querySelector('[data-option-wrapper]');
    const inputContainerDiv = document.querySelector('[data-select-container]');
    const clientHeight = document.documentElement.clientHeight;

    if (optionWrapperDiv && inputContainerDiv) {
      const boxOption = optionWrapperDiv.getBoundingClientRect();
      const boxInput = inputContainerDiv.getBoundingClientRect();

      const bottom = boxOption.top + boxOption.height;
      const bottomHeight = clientHeight - bottom;
      const bottomHeightInput =
        clientHeight -
        bottom -
        boxInput.height -
        parseInt(DROPDOWN_MARGIN_BOTTOM_TOP) -
        parseInt(error ? INPUT_BORDER_ERROR : INPUT_BORDER_ACTIVE) * 2;

      if (bottomHeight < 0) {
        setDropdownListMoveBorders(true);
      }
      if (bottomHeightInput > boxOption.height) {
        setDropdownListMoveBorders(false);
      }
    }

    // Показать / закрыть раскрывающийся список
    menuOpenAndClose ? setOpened(menuOpenAndClose) : setOpened((menuIsOpen: any) => !menuIsOpen);
  }, [setOpened, menuOpenAndClose]);

  const onBlur = (e: any) => {
    if (!e.currentTarget.contains(e.relatedTarget)) {
      setOpened(false);
    }
  };

  return (
    <SelectWrapper className={className} width={width} data-close-border>
      {label && (
        <Label disabled={disabled} menuIsOpen={menuIsOpen}>
          {label}
        </Label>
      )}
      <SelectContainer
        data-select-container
        disabled={disabled}
        menuIsOpen={menuIsOpen}
        size={size}
        type={type}
        error={error}
        dropdownListBorders={dropdownListBorders}
        tabIndex={1}
        onBlur={onBlur}
        onClick={handleSelectWrapperClick}
        className={classNameSelectContainer}
      >
        <Input
          size={size}
          disabled={disabled}
          placeholder={placeholder}
          clearable={clearable}
          menuIsOpen={menuIsOpen}
          currentMultiValue={value}
          maxWidthChip={maxWidthChip}
          onChange={onChange}
        />
        {!disabled && (
          <DropdownList
            size={size}
            disabled={disabled}
            menuIsOpen={menuIsOpen}
            error={error}
            optionsList={optionsList}
            dropdownListBorders={dropdownListBorders}
            currentMultiValue={value}
            setOpened={setOpened}
            onChange={onChange}
            classNameDropdownList={classNameDropdownList}
          />
        )}
      </SelectContainer>
      {additionalText && (
        <AdditionalText error={error} disabled={disabled} menuIsOpen={menuIsOpen}>
          {additionalText}
        </AdditionalText>
      )}
    </SelectWrapper>
  );
};
