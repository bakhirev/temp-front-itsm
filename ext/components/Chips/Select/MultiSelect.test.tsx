import React from 'react';
import { shallow } from 'enzyme';

import { IItemChipSelect } from './types';
import { MultiSelect } from './MultiSelect';

const listData: IItemChipSelect[] = [
  { id: '1', label: 'Суздаль' },
  { id: '2', label: 'Москва', disabled: true },
];

const listItemsDisabled: IItemChipSelect[] = [
  { id: 'a', label: 'Суздаль', disabled: true },
  { id: 'b', label: 'Москва', disabled: true },
  { id: 'c', label: 'Тверь', disabled: true },
  { id: 'd', label: 'Самара', disabled: true },
];

jest.mock('../../common', () => ({
  DEFAULT_THEME: {},
  Z_INDEXES: {
    TOOLTIP: 777,
  },
}));

describe('Chips.MultiSelect', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render component with items', () => {
    const wrapper = shallow(<MultiSelect items={listData} size={'big'} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with disabled items', () => {
    const wrapper = shallow(<MultiSelect items={listItemsDisabled} size={'big'} />);
    expect(wrapper).toMatchSnapshot();
  });
});
