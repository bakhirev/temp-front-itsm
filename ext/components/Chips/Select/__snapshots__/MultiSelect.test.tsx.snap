// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chips.MultiSelect should render component with disabled items 1`] = `
<styled.div>
  <styled.div
    disabled={true}
    focused={false}
    selected={false}
    size="big"
    tabIndex={-1}
    theme={Object {}}
    tooltip="Суздаль"
  >
    <styled.div
      size="big"
      withIcon={false}
    >
      <Styled(styled.div)>
        Суздаль
      </Styled(styled.div)>
    </styled.div>
  </styled.div>
  <styled.div />
  <styled.div
    disabled={true}
    focused={false}
    selected={false}
    size="big"
    tabIndex={-1}
    theme={Object {}}
    tooltip="Москва"
  >
    <styled.div
      size="big"
      withIcon={false}
    >
      <Styled(styled.div)>
        Москва
      </Styled(styled.div)>
    </styled.div>
  </styled.div>
  <styled.div />
  <styled.div
    disabled={true}
    focused={false}
    selected={false}
    size="big"
    tabIndex={-1}
    theme={Object {}}
    tooltip="Тверь"
  >
    <styled.div
      size="big"
      withIcon={false}
    >
      <Styled(styled.div)>
        Тверь
      </Styled(styled.div)>
    </styled.div>
  </styled.div>
  <styled.div />
  <styled.div
    disabled={true}
    focused={false}
    selected={false}
    size="big"
    tabIndex={-1}
    theme={Object {}}
    tooltip="Самара"
  >
    <styled.div
      size="big"
      withIcon={false}
    >
      <Styled(styled.div)>
        Самара
      </Styled(styled.div)>
    </styled.div>
  </styled.div>
</styled.div>
`;

exports[`Chips.MultiSelect should render component with items 1`] = `
<styled.div>
  <styled.div
    focused={false}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    selected={false}
    size="big"
    tabIndex={1}
    theme={Object {}}
    tooltip="Суздаль"
  >
    <styled.div
      size="big"
      withIcon={false}
    >
      <Styled(styled.div)>
        Суздаль
      </Styled(styled.div)>
    </styled.div>
  </styled.div>
  <styled.div />
  <styled.div
    disabled={true}
    focused={false}
    selected={false}
    size="big"
    tabIndex={-1}
    theme={Object {}}
    tooltip="Москва"
  >
    <styled.div
      size="big"
      withIcon={false}
    >
      <Styled(styled.div)>
        Москва
      </Styled(styled.div)>
    </styled.div>
  </styled.div>
</styled.div>
`;
